DEP: Pending deprecation warning for matrix

mhvk opened the pull request:
As discussed on the mailing list. ~Two commits, which can be split apart if need be~
* [X] make clearer in the docs that we do not recommend using `matrix`: done in #10973
* [X] emit `PendingDeprecationWarning` in `matrix.__new__`: this PR; UPDATE: includes silencing for the relevant tests.

The idea is only to warn users to start moving away; deprecation itself will depend on `scipy.sparse` as well as possible other packages.

Follow-up to this might be to move tests that are specific to matrix to `libmatrix` instead of scattered throughout the code, to help make a move to a separate package or other location easier.

p.s. Do let me know if I'm missing some obvious pieces of the documentation...

mhvk committed:
DEP: give PendingDeprecationWarning for use of the matrix subclass.

njsmith commented:
@mhvk can you give a sense of the slowdown the `PendingDeprecationWarning` gives for some small `matrix` operations? sometimes the `warning` module is surprisingly slow so I just want to double-check...

shoyer commented on numpy/matrixlib/defmatrix.py:
```diff
@@ -210,6 +215,9 @@ class matrix(N.ndarray):
     def __new__(subtype, data, dtype=None, copy=True):
+        warnings.warn('the matrix subclass is not the recommended way to deal '
+                      'with linear algebra. Please adjust your code to use '
+                      'regular ndarray.', PendingDeprecationWarning)
```
It would be a nice touch to add a URL pointing users to a longer description/discussion of why matrix is no longer recommended.

stefanv replied:
This may also confuse users who aren't attempting to do linear algebra, but simply thought that `np.matrix` is the default ndarray constructor.

ilayn replied:
Please also consider adding `stacklevel=1` to `warn()` such that the warning is raised from the calling function or `__main__` but not inside `np.matrix`.

shoyer reviewed (commented):
yes!

eric-wieser commented:
I think `np.array([1, 1, 2]).view(np.matrix)` skips `__new__`

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -309,8 +278,7 @@ Linear Algebra Equivalents
    * - ``[ a b; c d ]``
-     - ``vstack([hstack([a,b]), hstack([c,d])])`` or
-       ``bmat('a b; c d').A``
+     - ``block([[a,b], [c,d][)``
```
typo

mhvk commented:
@njsmith -on my slow laptop, `%timeit m = np.matrix([[0.,1.],[2.,3.]])` goes from 17.5 µs to 21 µs, so it is not horrendous.

@eric-wieser - I thought of doing it in `__array_finalize__`, but worried about making the impact too big: then every operation gets a 5 µs overhead.

@shoyer - Yes, a link would be better, which probably means that the note for the matrix subclass should be expanded to provide that description in more detail. I ran out of steam... (edits of or PRs to my PR most welcome...)

mhvk commented:
All tests with `matrix` now fail, which I guess means one should guard against the warning. One "advantage" is that it should make it easier to identify and move all those tests...

eric-wieser commented:
> but worried about making the impact too big: then every operation gets a 5 µs overhead.

You're right, `__new__` is exactly the right place, since that going to almost always be the user calling it.

> which I guess means one should guard against the warning. 

Perhaps only do so in the matrixlib test files, which will force you to move the tests?

njsmith commented:
We should check that whatever `scipy.sparse` is doing to create `matrix` objects triggers the warning, since we know that that's a major way to end up with `matrix` objects without manually using `np.matrix`.

stefanv commented:
SciPy usage:

```
compressed.py:        return np.matrix(result, copy=False)
compressed.py:            out = np.matrix(out)
coo.py:        return np.matrix(result, copy=False)
dia.py:            ret = np.matrix(res, dtype=res_dtype)
dia.py:            row_sums = np.matrix(row_sums)
dia.py:            ret = np.matrix(row_sums.sum(axis=axis))
```

```
base.py:                result = np.asmatrix(result)
base.py:                result = np.asmatrix(result)
base.py:        return np.asmatrix(self.toarray(order=order, out=out))
base.py:            return (self * np.asmatrix(np.ones(
base.py:            ret = np.asmatrix(np.ones(
base.py:            ret = self * np.asmatrix(
compressed.py:            ret = np.asmatrix(ret)
csr.py:            return np.asmatrix(val)
data.py:        return np.asmatrix(ret)
lil.py:                A = np.asmatrix(arg1)
linalg/interface.py:            y = np.asmatrix(y)
linalg/interface.py:            y = np.asmatrix(y)
linalg/interface.py:            Y = np.asmatrix(Y)
linalg/isolve/utils.py:            x = asmatrix(x)
```

stefanv commented on doc/source/reference/arrays.classes.rst:
```diff
@@ -215,6 +215,13 @@ Matrix objects
+.. note::
+   It is strongly advised *not* to use the matrix subclass.  As described
+   below, it makes writing functions that deal consistently with matrices
+   and regular arrays very difficult. Currently, the main reason still to use
```
This sentence is incomplete.

Perhaps "Currently, they are mainly used for interacting with `scipy.sparse`."

stefanv commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -162,35 +166,17 @@ There are pros and cons to using both:
       ``multiply(A,B)``.
    -  ``<:(`` The use of operator overloading is a bit illogical: ``*``
       does not work element-wise but ``/`` does.
+   -  You interact well with ``scipy.sparse``.
```
"You interact well" is kind of vague.  Same above.  Can we be more explicit here?

ahaldane commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -82,6 +82,8 @@ had to use ``dot`` instead of ``*`` to multiply (reduce) two tensors
 (scalar product, matrix vector multiplication etc.). Since Python 3.5 you
```
Above in the intro to this section, I would add in some of the description of `np.matrix` that you deleted above, otherwise the discussion to matrix seems out of the blue.

Maybe something like 
```
 'array' or 'matrix'? Which should I use?
========================================

Historically, NumPy has provided a special matrix type, `np.matrix`, which is a subclass of ndarray which makes binary operations linear algebra operations. You may see it used in some existing code. However, we recommend you do not use `np.matrix`, and instead use plain arrays for the following reasons:
```

stefanv commented:
@mhvk What is the status of this discussion?  Should the discussion be summarized in a short NEP, or should we simply move forward with a few PRs?

mhvk commented:
@stefanv - oops, that really slipped my mind. I think the sensible way here is to incorporate comments in the PR, so that there is something specific to discuss. I don't think this needs a NEP (we had a long discussion on the mailing list preceding this). As discussed at the top, we might still want to split the docs update (which is not controversial) at all from the `PendingDeprecationWarning`).

If you have time to take over the PR, please do feel free to use this or just start over.

mhvk commented:
Rebased just to see how error messages look (don't seem to get any test failures locally any more, perhaps because of the move to `pytest`?).  Note that this still includes the `DOC` commit, which I split off to #10973; I'll rebase if that is merged soon (or can remove it here if that helps).

charris commented:
I think we can ignore that warning in `PytestTester` and `pytest.ini`. Adding
```
# Ignore matrix PendingDeprecationWarning
    ignore:the matrix subclass is not:PendingDeprecationWarning
```
to `pytest.ini` fixes the development tests, and adding
```
            "-W ignore:the matrix subclass is not:PendingDeprecationWarning",
```
to `PytestTester` should take care of the release.

Might be able to use module level setup and teardown also, but I don't know that we can trust the warnings module in all the python versions we support.

charris commented:
I don't know what to do about SciPy, maybe put guards around all the matrix creations? @pv @rgommers thoughts?

mhvk committed:
TST: Ignore PendingDeprecationWarning in matrixlib tests.

mhvk commented:
@charris - OK, I rebased and ignore the warnings a bit more elegantly, by specifically silencing them only in the `matrixlib` tests (in every test file; not completely sure why I cannot seem to do it for the whole module).

While doing that, I also noticed that I had missed a few tests in `np.ma`; see #11137 (on which this is based; first commit).

mhvk commented:
p.s. Fine to postpone this to 1.16; we want to get this right. But probably good to get #11137 in, so in 1.15 at least all the tests have been moved.

mhvk commented:
Hmm, anybody have a clue why the `chroot/artful` travis test failed to pick up my warning filter. Pytest too old perhaps?

mattip commented:
The arch job seems to be picking up the ini file and  [pytest 3.1.3](https://travis-ci.org/numpy/numpy/jobs/382347591#L4883) from the [system install](https://travis-ci.org/numpy/numpy/jobs/382347591#L1274)

mattip commented:
.. and indeed filterwarnings was added in commit 7341da1 which is contained in versions 3.2.0 and up.

mhvk commented:
@mattip - nice sleuthing! Am not completely sure what to do, though. Do we adjust that test? Or should I try to add a work-around for older pytest?

mattip commented:
why are we pinning the 1386 build to artful which is 17.10, when 18.04 [has been released](https://wiki.ubuntu.com/Releases)? bionic has [pytest 3.3](https://packages.ubuntu.com/bionic/python/pypy-pytest)

mattip commented:
It seems it is time to update PR #10425 which moved from zesty to artful, and has a comment that we should update to bionic when it is released

mattip commented:
PR #11140 should be enough to fix the failing i386 test run

rgommers commented:
> I don't know what to do about SciPy, maybe put guards around all the matrix creations? @pv @rgommers thoughts?

Too much trouble I'd say. I'd just add an ignore filter for this warning in `scipy/sparse/__init__.py`.

rgommers commented on numpy/matrixlib/defmatrix.py:
```diff
@@ -70,6 +71,10 @@ class matrix(N.ndarray):
 
+    .. note:: It is no longer recommended to use this class, even for linear
+              algebra. Instead use regular arrays. The class may be removed
+              in the future.
```
If you mean https://github.com/numpy/numpy/pull/10142#issuecomment-391548138, then I don't think it does (at least, it doesn't matter how the `PendingDeprecationWarning` is silenced). `PendingDeprecationWarning`s also won't show up for normal users, only if you run with `-Wd` or raise exceptions on warnings or something like that.

+1 for getting this note into 1.15

charris commented:
Thanks Marten.

@rgommer NumPy will now raise the `PendingDeprecationWarning`, do what you will with SciPy.

mhvk commented:
:fireworks:

charris commented:
Just to note that I needed pytest 3.6.1, 3.5 was failing.
